Update holdings records, delete old holdings

jsteverman opened the pull request:
Adds the option to update holdings with bin/add_print_holdings.rb

ClusterHolding.new(h).update finds a matching holdings record and updates its received_date, or adds it if none can be found. For our purposes matching means all fields are identical except date_received. This overrides Holding's == method. 

After updating holdings records, bin/add_print_holdings.rb deletes any remaining holdings from the particular organization and an older date_received. 

While testing it was discovered the Cluster.ocns index was not being used resulting in full column scans. The partial filter has been changed from "ocns.0" : {$exists:1} to ocns :{$gt:1}

aelkiss commented:
I think this will work and makes sense as an approach when we're receiving both incremental updates and full updates.

I wonder if as an optimization for the current situation in which members can only submit full dumps but do so frequently if we could avoid having to touch every record by instead:

- computing a delta (e.g. using comm, a similar approach to this: https://github.com/hathitrust/print_holdings_redux/blob/master/get_delta/comm_concordance_delta.sh)

- using `ClusterHolding.new(h).delete` for any lines in the old file that are not in the new file
- using `ClusterHolding.new(h).cluster ` for any lines in the new file that are not in the old file

Anything new would just get added; anything updated would get deleted and then re-added with the new attributes; anything missing would just get deleted.

In a world where we receive both full updates and incremental updates (in some format) it would 1) become much more difficult to compute this delta, since any incremental updates since the last full update would have to be taken into account, and 2) (hopefully) much rarer to receive a full update.

Thoughts?

billdueber commented:
Computing/storing the deltas will be kind of expensive, but nowhere near as expensive as blindly deleting/re-adding everything. The cleanest way I can think of would be to have, for each partner, a giant table of triples of the form `oclc, line-hash, count` that we'd be continually updating. A partial load would have to be split by the patron into deletes/adds, so that's an easy update. A full load would essentially force us to load up a temp table and do queries against the two versions to get our delete/add lists. 

I know we're talking about hundreds of millions of lines, which is kinda exactly why I would want to try to deal with it in a database and not as a directory of text files.

jsteverman commented:
Rebased. 

While taking the approach found in this PR, I think I addressed the conceptual problems I was having with ```sort | comm```. I am still wary of relying on the shuffling of thousands of tsv files and how it work with incremental updates. 

I am currently running an add of all UMich monos (5.4m) from last year and then an update of all UMich monos from this year, and will report on performance tomorrow.

billdueber commented:
Adding comment from @jsteverman from slack:

>So it took 4 hours and 15 minutes to load all of UMich's 2019 monographs (5.4m). It took another 3 hours and 45 minutes to update with the 2020 monographs.

billdueber commented on lib/cluster_holding.rb:
```diff
@@ -24,4 +24,41 @@ def move(new_cluster)
+
+    old_holding = c.holdings.to_a.find do |h|
+      h == @holding &&
+        h.date_received != @holding.date_received
```
Do you think it's worth defining `#===` to mean "equal including the date"? Might be nice if we extend which attributes count/don't count.

billdueber commented on bin/add_print_holdings.rb:
```diff
@@ -24,13 +25,43 @@
+  current_date = (current_date || h.date_received)
+  organization = (organization || h.organization)
+  c = if update
+    ClusterHolding.new(h).update
```
Probably premature optimization, but  it looks like the ClusterHolding service looks up the cluster based on the holding's OCN, but at least in this case we've got the cluster in our hot little hands from line 31. No idea how expensive that lookup is, but might be worth thinking about being able to optionally pass in the cluster at ClusterHolding initialization (unless we want to look it up every time in case something changed underneath?)

jsteverman replied:
I think the line numbers have changed since you commented? 

This might be applicable to the delete operation in the update block starting at line 56

billdueber replied:
Right. In general, ClusterHolding takes a holding and the first thing it does for some operations is look up a cluster from the holding. If finding the cluster of a holding is at all expensive, we might want to allow a cluster passing in. 

We should probably have a JIRA ticket/super-ticket for benchmarking / profiling stuff.

aelkiss commented on lib/cluster_holding.rb:
```diff
@@ -24,4 +24,41 @@ def move(new_cluster)
+  end
+
+  # Removes a holding
+  def delete
+    c = Cluster.find_by(ocns: @holding.ocn)
+    @holding.delete
+    c.save
+    c.reload
+    c.delete unless c._children.any?
+  end
```
The delete should only happen if we deleted the last child, so I'm not sure how a race condition would happen. Happy to entertain a sequence of events that could cause one, though.

aelkiss replied:
Hmm, that could happen, although I agree it's very unlikely. I think there is an atomic mongodb `delete_if` operation that might be helpful here? I'm also not sure how `_children` works, so I'm not sure this will ever actually return true - would the holding ever be the only thing in `_children`? I also wonder if we have the same race condition concern for all the split/merge operations - let's say you were in the middle of doing a merge operation that caused one cluster to be deleted, but another thread already had a reference to that to-be-deleted cluster. Even if you could lock that cluster for updates (and I need to check whether that's possible in MongoDB) I think what would happen is that the cluster would get deleted but then in the other thread save would resurrect it? Maybe there's a way to have save raise an exception if you try to save something that got deleted in the meantime, and then it could just retry that operation? That seems like a potentially frequent concern, in any case. Will have to investigate.

jsteverman replied:
are we worried about a race condition here?

Always, but I'm still working on an understanding of multithreading with Mongoid. https://docs.mongodb.com/mongoid/current/tutorials/mongoid-transactions/

._children are a document's embedded documents. In the case of a Cluster, that's anything other than the ocns field. I suspect clusters will frequently consist of a single ocn and the holding record it is taken from since we have 400m of them. Looking at it again, this functionally identical to Reclusterer.new(c).recluster

billdueber commented:
If multiple threads are working on different institutions, couldn't one be
adding a holding as the other is deleting the record? Like I said, it's not
exactly a _high_ probability...

On Tue, Jun 30, 2020 at 12:18 PM Aaron Elkiss <notifications@github.com>
wrote:

> *@aelkiss* commented on this pull request.
> ------------------------------
>
> In lib/cluster_holding.rb
> <https://github.com/hathitrust/holdings-backend/pull/28#discussion_r447810202>
> :
>
> > +  def delete
> +    c = Cluster.find_by(ocns: @holding.ocn)
> +    @holding.delete
> +    c.save
> +    c.reload
> +    c.delete unless c._children.any?
> +  end
>
> The delete should only happen if we deleted the last child, so I'm not
> sure how a race condition would happen. Happy to entertain a sequence of
> events that could cause one, though.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/hathitrust/holdings-backend/pull/28#discussion_r447810202>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAA32VQ26NB73WA53YOJ3ZDRZIF5ZANCNFSM4OLNO4KQ>
> .
>
-- 
Bill Dueber
Library Systems Programmer
University of Michigan Library

jsteverman commented on bin/add_print_holdings.rb:
```diff
@@ -24,13 +25,43 @@
+if update
+  Cluster.where("holdings.organization": organization,
+                "holdings.date_received": { "$lt": current_date }).each do |c|
+                  c.holdings.to_a.each do |h|
```
This was actually the last 10% that accounted for the other 90% of the effort. Naturally some of the clean up I did at the end broke the test that proved it.

Embarrassingly, I haven't nailed down the exact mechanism at work, but deleting within the cursor borks the cursor.  One of the effects is that clusters with multiple holdings needing deletion might not make through all of the holdings. 

I need to rewrite/add a test, and should be using ClusterHolding.delete_old_holdings.
